### PR TITLE
64bit memory calculations

### PR DESCRIPTION
**Pull request by holiman**
This PR replaces https://github.com/ethereum/go-ethereum/pull/19203 . 

This PR contains five separate optimizations to the evm runloop. For review, it's probably simplest to verify each commit separately, since they are completely standalone, and some larger than others. For those, I've tried to make them as easy to verify as possible, by making the changes verifiable side-by-side with the original code. 

The changes are:

## 1. Remove function call for stack validation

Previously, we always called `operation.validateStack`. Since `operation.validateStack` is a pointer, this call cannot be inlined by the compiler. It has been replaced with `minStack` and `maxStack` on the `operation`, which can be checked immediately in the runloop. 

Additionally, I removed the call to `enforceRestrictions`, by inlining it. 

## 2. Separate gas calculation into constant and dynamic

Previously, we always called `operation.gasCost(in.gasTable, in.evm, contract, stack, mem, memorySize)` (again, cannot be inlined). 

For a lot of operations, this method only returned a constant. I instead added the constant gas values to those operations, which can be checked immediately. This makes a difference particularly for all cheap operations which have no dynamic part, which are the majority of the operations. 

## 3 Optimize push1

This adds a custom `opPush1` method for push1, where we can convert a byte directly into a `uint64`, and don't have to do padding which the generic method does. 

## 4 Reuse pooled bigints for `ADDRESS`, `ORIGIN`, and `CALLER`

This adds a convenience-method to `Address` to convert into an existing bigint, and the ops above can thus use the int pool instead of allocating new bigints. 

## 5 Avoid string interpolation for jump errors

On a trial run with metrics, it was found that the second largest time-sink was `7m30s` in JUMPI (`1m31s` for `JUMP`).  Now, since solidity used to do intentional invalid jumps for abort all the time, most of that time is probably spent on `fmt.Errorf("invalid jump destination (%v) %v", nop, pos)`. 

This PR replaces the string interpolation with a generic `errInvalidJump` error. 

## 6 Avoid allocating bigints during memory calc

It changes the signature of the memory calculation function to use `uint64` instead of bigints. 

WIthout this PR, 

* A new bigint was always allocated when the memory size was calculated (except if length was zero). 
* In some cases (`MSTORE8`, `MLOAD`, `MSTORE`), and additional bigint was allocated since the `length` does not come from the stack. 

As a side-effect, overflows are now detected and aborted earlier.

### CODE REVIEW

**Comment by holiman** [comment]
Benchmarks after ~24hours of full sync. This PR is roughly 1 hour ahead.

screen shot 2019-03-06 at 09 57 03

**Comment by holiman** [comment]
After roughly 48 hours, this PR is about 2 hours ahead
screen shot 2019-03-07 at 08 15 38

**Comment by holiman** [comment]
After 3 days (and two hours)
screen shot 2019-03-08 at 12 43 07
THe PR is about 3 hours ahead: 
screen shot 2019-03-08 at 12 44 10

**Comment by holiman** [comment]
After almost 4.5 days, it's almost 4.5 hours ahead. I actually would have thought that the linearity would break on the disk-heavy sections (5-6M and later), but that does not appear to be the case. 
Screen Shot 2019-03-09 at 20 16 43

So it's about one hours every 24hours, or ~4 percent performance increase.

**Comment by holiman** [comment]
They both ran out of disk before completion, but made it to `7.13M` and `7.11M` respectively. 
`mon12` at `7118879` at `Mar 10 07:23:46`
`mon13` at `7118975` at `Mar 10 11:43:03`, `4h 40m`

Screen Shot 2019-03-10 at 21 21 10

File: core/vm/common.go [hunk 1]
```diff
@@ -23,13 +23,47 @@ import (
...
+// calcMemSize64 calculates the required memory size, and returns
+// the size and whether the result overflowed uint64
+func calcMemSize64(off, l *big.Int) (uint64, bool) {
+	// if length is zero, memsize is always zero, regardless of offset
 	if l.Sign() == 0 {
-		return common.Big0
+		return 0, true
 	}
+	// Check that neither offset nor length overflows
+	if off.BitLen() > 64 || l.BitLen() > 64 {
```
Comment (fjl) [inline core/vm/common.go]: use `IsUint64`.

File: core/vm/common.go [hunk 2]
```diff
@@ -23,13 +23,47 @@ import (
 	"github.com/ethereum/go-ethereum/common/math"
 )
 
-// calculates the memory size required for a step
-func calcMemSize(off, l *big.Int) *big.Int {
+// calcMemSize64 calculates the required memory size, and returns
+// the size and whether the result overflowed uint64
+func calcMemSize64(off, l *big.Int) (uint64, bool) {
```
Comment (fjl) [inline core/vm/common.go]:
```
if !l.IsUint64() {
    return 0, true
}
calcMemSizeWithUint(off, l.Uint64())
```

File: core/vm/common.go [hunk 3]
```diff
@@ -23,13 +23,47 @@ import (
...
+// calcMemSize64WithUint calculates the required memory size, and returns
+// the size and whether the result overflowed uint64
+// Identical to calcMemSize64, but length is a uint64
+func calcMemSize64WithUint(off *big.Int, length64 uint64) (uint64, bool) {
+	// if length is zero, memsize is always zero, regardless of offset
+	if length64 == 0 {
+		return 0, true
+	}
+	// Check that offset doesn't overflow
+	if off.BitLen() > 64 {
```
Comment (fjl) [inline core/vm/common.go]: `IsUint64()`

File: core/vm/common.go [hunk 4]
```diff
@@ -23,13 +23,47 @@ import (
...
+		return 0, true
+	}
+	offset64 := off.Uint64()
+	val := offset64 + length64
+	// Check that the total doesn't overflow
+	if val < offset64 {
+		return 0, true
+	}
+	return val, false
```
Comment (fjl) [inline core/vm/common.go]: `return val, val < offset64`

**Comment by holiman** [comment]
Updated, PTAL

File: core/vm/interpreter.go [hunk 5]
```diff
@@ -217,19 +201,35 @@ func (in *EVMInterpreter) Run(contract *Contract, input []byte, readOnly bool) (
...
 		// If the operation is valid, enforce and write restrictions
-		if err = in.enforceRestrictions(op, operation, stack); err != nil {
-			return nil, err
+		if in.readOnly && in.evm.chainRules.IsByzantium {
+			// If the interpreter is operating in readonly mode, make sure no
+			// state-modifying operation is performed. The 3rd stack item
+			// for a call operation is the value. Transferring value from one
+			// account to the others means the state is modified and should also
+			// return with an error.
+			if operation.writes || (op == CALL && stack.Back(2).BitLen() > 0) {
```
Comment (karalabe) [inline core/vm/interpreter.go]: Could we use `stack.Back(2).Sign` here instead of bitlen?

**Comment by holiman** [comment]
Updated, I removed a couple of occurences where `BitLen` could be replaced by `Sign` or `IsUint64`

**Pull request merged**